### PR TITLE
Add new circuitbreaker plugin and refactor CircuitBreakerService

### PR DESCRIPTION
**Pull request by benwtrent**
This commit lays the ground work for plugins supplying their own circuit breakers. 

It adds a new interface: `CircuitBreakerPlugin`.

This interface provides methods for providing custom child CircuitBreaker objects. There are also facilities for allowing dynamic settings for the custom breakers.

With the refactor, circuit breakers are no longer replaced on setting changes. Instead, the two mutable settings themselves are `volatile`. Plugins that want to use their custom circuit breaker should keep a reference of their constructed breaker.

### CODE REVIEW

**Comment by benwtrent** [comment]
//CC @elastic/ml-core 

This PR lays the ground work for custom circuit breakers.

File: server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import org.elasticsearch.indices.breaker.CircuitBreakerService;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * An extension point for {@link Plugin} implementations to add custom circuit breakers
+ */
+public interface CircuitMemoryBreakerPlugin {
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java:33]: Drive-by comment -- would this be better named `CircuitBreakerPlugin`?

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java:33]: For sure :)

File: server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    default List<BreakerSettings> getCircuitBreakers() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * If the custom {@link BreakerSettings} have dynamic settings.
+     * A settings update consumer can be added with this method.
+     * Once the new settings are constructed, pass them to the updatedBreakerSettingsListener.
+     */
+    default void addDynamicBreakerUpdates(ClusterSettings settings, Consumer<BreakerSettings> updatedBreakerSettingsListener) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java:51]: All of our plugin apis should be "pull" based, where we extract from the plugin what will be added to ES. This seems to do the opposite. It's unclear what from my reading of this PR what this method actual does, since there are no example or test implementations, but based on the one use inside Node, it seems like this is used as a way to callback to actually add the breaker to the CircuitBreakerService. I think the root of the problem is CircuitBreakerService should not have a registerBreaker method. Instead, Node should collect all the breaker settings, and pass them to the service's ctor, as we don't need the ability to add these dynamically.

Additionally, the BreakerSettings concept seems like it needs a little work to make this all truly extendable. Currently it takes in all the values, but the plugin must read these values on its own, and the `getCircuitBreakers()` method above then relies on the plugin having Settings from its ctor, which cannot be relied on. Even if we passed the real Settings to `getCircuitBreakers()`, it seems opposite of what we want. IMO any new breakers that are added should be configurable in the same way our existing breakers are configured, ie through the same yaml settings with the new breaker name substituted.  This would mean replacing the hardcoded settings that exist for the builtin breakers with affix settings, so that they can be used with any breaker.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java:51]: > It's unclear from my reading of this PR what this method actually does, since there are no example or test implementations, but based on the one use inside Node, it seems like this is used as a way to callback to actually add the breaker to the CircuitBreakerService

It is a way to hook into dynamic setting updates for the new circuit breakers. If this is possible by have a settings listener on a suffix, that would definitely be better. 

> IMO any new breakers that are added should be configurable in the same way our existing breakers are configured, ie through the same yaml settings with the new breaker name substituted.

Effectively, new circuit breakers would be added by plugins declaring new settings that match the appropriate affix?

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: @rjernst I have updated the setting handling. Now new breakers are affix settings. The initial breaker settings are still provided by the plugin, but facilities are in place to make that as seamless as possible.

Still an open question around making the current circuit breakers affix settings and deprecating their current setting values.

Also, I opted to NOT make `durability` configurable. This seems like something a user should NEVER be able to change (even at start up time).

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java (hunk 2 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: It still seems like there is more boiler plate than there should be. What is the plugin actually providing? The setting parsing/registration should be possible within core, where plugins never need to do anything to handle it. Instead, plugins can just provide the breaker name, and core registers the appropriate settings/update handlers. All the plugin needs is a reference to the CircuiteBreaker so it can make calls to add/remove allocations. Currently this is done by exposing the CircuitBreakerService, which has a get method based on the breaker name. This isn't the worst for now, but the registerBreaker method should go away. It was added in a time when we were still fully utilizing guice, instead of explicit construction and pull based plugin apis that we have now.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: > All the plugin needs is a reference to the CircuitBreaker so it can make calls to add/remove allocations. 

This is not true. The underlying references change when there are setting changes. The plugin won't lose the data it was accounting, but it would not see setting updates until a reboot. Which effectively makes plugin breaker settings static.

> The setting parsing/registration should be possible within core, where plugins never need to do anything to handle it. Instead, plugins can just provide the breaker name, and core registers the appropriate settings/update handlers. 

How can different plugins provide their appropriate setting defaults? To my knowledge, affix settings have global defaults and different plugins might want to provide their own sane defaults.

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: > The underlying references change when there are setting changes. 

This seems pretty bad. There are ways to make the writing use locking but the read side, which is more common, not slow down. Also, after looking at the register method again, I've noticed the comment noting registering the same name replaces the existing breaker. This is leniency which I think makes having this available to plugins error prone. What if two plugins provide the same breaker name? Then the plugin registration order matters, and one  silently replaces the other. This is yet another reason I think we need to statically define the breakers at node startup, but still have them available.

> How can different plugins provide their appropriate setting defaults?

This is a good point and something I hadn't considered. There are ways around this though. For example, in ScriptService we now use affix settings for context specific cache values, where the defaults for those are defined in the contexts themselves (ie like circuit breakers types here). It means the default defined in the setting is ignored and we apply the defaults dynamically based on whether the setting exists, rather than calling `setting.get` always.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: > There are ways to make the writing use locking but the read side, which is more common, not slow down. 

Keeping the reference the same across threads will require overhead & limit to be `volatile` on read no? How else will threads ever see the updates? I think this will have a non-negligible performance hit. We might get around this hit by making method local copies of the volatile fields so they are only read once per method call.

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: What we have done in other places is wrap the state into a single internal object (eg look at XPackLicenseState.Status). Updates to the setttings construct a new object, and flip all the values at once through a single `volatile`.

> I think this will have a non-negligible performance hit

This is the pattern we have used in numerous places in Elasticsearch. I know of no other place where we poll a service for the current instance of an object on every call. Having services/objects update themselves through cluster state listeners and volatile state is very common, and not a performance hit in any cases I know of.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/plugins/CircuitMemoryBreakerPlugin.java]: @rjernst volatile reads are 3x as slow as non-volatile. But, given the other changes, I think it will be OK (we won't need concurrent hash map any longer).

I am going to work on implementing this (initializing via ctor, volatile settings, etc.) and test performance with JMH. Will get back once done. 

Thanks so much for your thoughts and time on this!

File: benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.concurrent.TimeUnit;
+
+@Fork(3)
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Benchmark)
+@SuppressWarnings("unused") // invoked by benchmarking framework
+public class HierarchyCircuitBreakerBenchmark {
```
Comment (benwtrent) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java:50]: This can be deleted before merge. I am using it for micro-benchmarking the change as circuit breakers are hot spots in the code base.

**Comment by benwtrent** [comment]
Micro-benchmark results:

master:
```
Benchmark                                                               (breakerNames)  Mode  Cnt     Score    Error  Units
HierarchyCircuitBreakerBenchmark.circuitBreak_1                                request  avgt   15     8.114 ±  0.287  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_1   request,inflight_requests,accounting  avgt   15    24.191 ±  1.136  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_4                                request  avgt   15    42.657 ±  1.835  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_4   request,inflight_requests,accounting  avgt   15   102.847 ± 11.784  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_64                               request  avgt   15   759.323 ± 18.518  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_64  request,inflight_requests,accounting  avgt   15  1106.358 ± 88.316  us/op
```
With changes:
```
Benchmark                                                               (breakerNames)  Mode  Cnt     Score     Error  Units
HierarchyCircuitBreakerBenchmark.circuitBreak_1                                request  avgt   15     5.922 ±   0.286  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_1   request,inflight_requests,accounting  avgt   15    17.347 ±   0.206  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_4                                request  avgt   15    43.332 ±   1.413  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_4   request,inflight_requests,accounting  avgt   15    76.662 ±  10.156  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_64                               request  avgt   15   808.669 ±  39.591  us/op
HierarchyCircuitBreakerBenchmark.circuitBreak_64  request,inflight_requests,accounting  avgt   15  1180.433 ± 105.286  us/op
```
So, either my benchmarks are useless, or the changes make no significant difference on performance for the circuit breaker service. 

I focused on the "non-breaking" and "non-update" scenario. The most common path should be adding and removing items from a circuit breaker whose settings are not being updated. Both setting updates and circuit breaks should be rare events.

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java [hunk 4]
```diff
@@ -331,56 +327,29 @@ public void checkParentLimit(long newBytesReserved, String label) throws Circuit
...
         }
     }
 
-    /**
-     * Allows to register a custom circuit breaker.
-     * Warning: Will overwrite any existing custom breaker with the same name.
-     */
-    @Override
-    public void registerBreaker(BreakerSettings breakerSettings) {
+    public CircuitBreaker validateAndCreateBreaker(BreakerSettings breakerSettings) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java:345]: `ChildMemoryCircuitBreaker` objects need a reference to the constructed `HierarchyCircuitBreakerService`. This complicates plugin access to their own created circuit breakers. To prevent muddying the API between the plugin and the node, I have split the breaker creation and the breaker registration.

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java [hunk 5]
```diff
@@ -158,57 +151,60 @@ public HierarchyCircuitBreakerService(Settings settings, ClusterSettings cluster
             this::setRequestBreakerLimit);
         clusterSettings.addSettingsUpdateConsumer(ACCOUNTING_CIRCUIT_BREAKER_LIMIT_SETTING, ACCOUNTING_CIRCUIT_BREAKER_OVERHEAD_SETTING,
             this::setAccountingBreakerLimit);
+        clusterSettings.addAffixUpdateConsumer(CIRCUIT_BREAKER_LIMIT_SETTING, CIRCUIT_BREAKER_OVERHEAD_SETTING,
+            this::setCircuitBreakerLimit,
+            (s, t) -> {});
+    }
+
+    @Override
+    public void registerNewCircuitBreakers(List<CircuitBreaker> circuitBreakers) {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java:160]: For plugins to create their own circuit breaker instances (so that they can maintain a reference to them without the service), we still need to register circuit breakers outside of the constructor. 

Main difference here is that registration no longer overwrites previous circuit breakers and will throw if one is already registered with the same name.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java (hunk 3 above)
Comment (benwtrent) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java:50]: @rjernst would you like these benchmarks deleted for now? Or rather keep them in the repo for future use?

File: server/src/main/java/org/elasticsearch/plugins/CircuitBreakerPlugin.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+     * Custom circuit breakers settings can be found in {@link BreakerSettings}.
+     * See:
+     *  - limit (example: `breaker.foo.limit`) {@link BreakerSettings#CIRCUIT_BREAKER_LIMIT_SETTING}
+     *  - overhead (example: `breaker.foo.overhead`) {@link BreakerSettings#CIRCUIT_BREAKER_OVERHEAD_SETTING}
+     *  - type (example: `breaker.foo.type`) {@link BreakerSettings#CIRCUIT_BREAKER_TYPE}
+     *
+     * The `limit` and `overhead` settings will be dynamically updated in the circuit breaker service iff a {@link BreakerSettings}
+     * object with the same name is provided at node startup.
+     */
+    BreakerSettings getCircuitBreaker(Settings settings);
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/plugins/CircuitBreakerPlugin.java:47]: I opted to have this be singular + return a `BreakerSettings` object. A redirection through a factory does not really provide value, especially since the factory will almost be exactly like this method definition.

**Comment by benwtrent** [comment]
run elasticsearch-ci/default-distro

File: benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java (hunk 3 above)
Comment (rjernst) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java:50]: We should delete them. If they aren't being run regularly then it is only a nuisance to refactorings where we must change code that isn't actually maintained.

Comment (benwtrent) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/indices/breaker/HierarchyCircuitBreakerBenchmark.java:50]: Roger, I will delete :D.

File: server/src/main/java/org/elasticsearch/common/breaker/ChildMemoryCircuitBreaker.java [hunk 7]
```diff
@@ -208,15 +185,15 @@ public long getUsed() {
      */
     @Override
     public long getLimit() {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/breaker/ChildMemoryCircuitBreaker.java:187]: It looks like the one use of this is to grab stats for a breaker. Due to this and overhead being separate methods, this could theoretically result in a race condition where we get incorrect configuration in the stats. I realize this is already currently an issue, so this doesn't need to be solved right now, but could you please create a followup issue to solve this, maybe by getting rid of these methods and having a stats() method directly on the breaker.

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java [hunk 8]
```diff
@@ -95,120 +100,97 @@
...
     private volatile BreakerSettings parentSettings;
-    private volatile BreakerSettings fielddataSettings;
-    private volatile BreakerSettings inFlightRequestsSettings;
-    private volatile BreakerSettings requestSettings;
-    private volatile BreakerSettings accountingSettings;
 
     // Tripped count for when redistribution was attempted but wasn't successful
     private final AtomicLong parentTripCount = new AtomicLong(0);
 
     public HierarchyCircuitBreakerService(Settings settings, ClusterSettings clusterSettings) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java:107]: Why do we need this alternate ctor? Can we convert the few tests calling tihs to pass an empty list?

**Review by rjernst — APPROVED** [review]
Thanks for all the iterations here. I just left a few nits. LGTM

File: server/src/main/java/org/elasticsearch/indices/breaker/BreakerSettings.java [hunk 9]
```diff
@@ -69,7 +129,7 @@ public double getOverhead() {
     public String toString() {
         return "[" + this.name +
                 ",type=" + this.type.toString() +
-                ",durability=" + this.durability.toString() +
+                ",durability=" + (this.durability == null ? "null" : this.durability.toString()) +
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/indices/breaker/BreakerSettings.java:132]: @rjernst when the breaker settings are for the parent breaker.

File: server/src/main/java/org/elasticsearch/common/breaker/ChildMemoryCircuitBreaker.java (hunk 7 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/common/breaker/ChildMemoryCircuitBreaker.java:187]: @rjernst definitely, I will create a new ES issue. It does not make sense to have separate methods any longer.
